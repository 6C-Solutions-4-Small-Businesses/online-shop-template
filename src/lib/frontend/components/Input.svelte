<script lang="ts">
    export let id: string
    export let classNames: string = 'w-full'
    export let label = ''
    export let type = 'text'
    export let placeholder = ''
    export let value: string
    export let error = false
    export let disabled = false
    export let required = false
    export let hidden = false
    export let accept = '*'
    export let capture: boolean | 'user' | 'environment' | null | undefined = undefined
    export let step: string | undefined = undefined
    export let onChange: (event: Event) => void = () => {
    }
    export let onBlur: (event: Event) => void = () => {
    }

</script>

<div class="{classNames}" data-testid="input-container">
    <h4 class="mb-1 text-black" data-testid="input-label-id">{label}
        {#if required}
            <span class="text-red-500">*</span>
        {/if}
    </h4>
    <input
            class={`w-full h-fit ${hidden? 'hidden' : ''} p-3.5 rounded-md ${error ? 'border-2 border-red-500' : 'border-gray-500'} ${disabled ? 'bg-gray-100' : 'bg-white'} focus:outline-none focus:border-blue-500`}
            data-testid={id}
            disabled={disabled}
            id={id}
            on:input={onChange}
            placeholder={placeholder}
            accept="{accept}"
            type={type}
            value={value === undefined ? '' : value}
            capture="{capture}"
            on:blur={onBlur}
            step="{step}"
    />
</div>
