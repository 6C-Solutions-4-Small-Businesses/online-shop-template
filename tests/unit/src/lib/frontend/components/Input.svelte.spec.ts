import {cleanup, fireEvent, render, type RenderResult, screen} from '@testing-library/svelte'
import Input from '$lib/frontend/components/Input.svelte'
import '@testing-library/jest-dom'
import {afterEach, expect} from 'vitest'

describe('Input Component', () => {
    const inputId = 'test-id'
    const inputLabel = 'input-label'
    const inputPlaceHolder = 'input-placeholder'
    const initialInputValue = 'Initial Value'
    const inputType = 'text'
    const classNames = 'bg-red-500'
    const inputCaptureType = 'user'
    const onBlur = vi.fn()
    const onChange = vi.fn()

    let inputElement: HTMLInputElement
    let rendering: RenderResult<Input, typeof import('@testing-library/dom/types/queries')>

    beforeEach(() => {
        rendering = render(Input, {
            props: {
                id: inputId,
                label: inputLabel,
                type: inputType,
                placeholder: inputPlaceHolder,
                value: initialInputValue,
                classNames,
                capture: inputCaptureType,
                onBlur,
                onChange,
            },
        })
        inputElement = rendering.getByTestId(inputId) as HTMLInputElement
    })

    describe('html structure', () => {

        it('should have match snapshot', () => {
            expect(rendering.container).toMatchSnapshot()
        })

        it('should map the "id" property to the "data-testid" of the input', () => {

            expect(screen.getByTestId(inputId)).toBeInTheDocument()
        })

        it('should set the value of the label element to the value of the label property', () => {

            const labelElement = rendering.getByText(inputLabel)

            expect(labelElement).toBeInTheDocument()
        })

        it('should set the value of the input type to the value of the type property', () => {

            expect(inputElement.type).toBeDefined()
            expect(inputElement.type).toStrictEqual(inputType)
        })

        it('should set the value of the input placeholder to the value of the placeholder property', () => {

            expect(inputElement.placeholder).toBeDefined()
            expect(inputElement.placeholder).toStrictEqual(inputPlaceHolder)
        })

        it('should set the value of the input element to the initial value of the value property', () => {

            expect(inputElement.value).toBeDefined()
            expect(inputElement.value).toStrictEqual(initialInputValue)
        })

        it('should set the input disabled property to the value of the provided disabled property', () => {

            expect(inputElement.disabled).toBeDefined()
            expect(inputElement.disabled).toBeFalsy()
        })

        it('should set the input div container class to the provided "classNames" property', () => {

            const containerElement = rendering.getByTestId('input-container') as HTMLInputElement

            expect(containerElement.classList).toContain(classNames)
        })

        it('should set the hide the input if the "hidden" property is set to true', async () => {

            await rerenderComponent({hidden: true})

            const hiddenInputElement = rendering.getByTestId(inputId) as HTMLInputElement

            expect(hiddenInputElement.classList).toContain('hidden')
        })

        it('should set the "accept" property of the input to the value provided to the property "accept"', async () => {
            const accept = 'image/*'

            await rerenderComponent({accept})

            const imageInputElement = rendering.getByTestId(inputId) as HTMLInputElement
            expect(imageInputElement.accept).toStrictEqual(accept)
        })

        it('should set the increment and decrement step of the input to the value provided to the property "step"', async () => {
            const step = '0.0001'

            await rerenderComponent({step})

            const stepInputElement = rendering.getByTestId(inputId) as HTMLInputElement
            expect(stepInputElement.step).toStrictEqual(step)
        })
    })

    describe('Behavior', () => {

        it('should call onChange handler when a change occurs in the input', () => {

            fireEvent.input(inputElement, {target: {value: 'new value'}})

            expect(onChange).toHaveBeenCalled()
        })

        it('should call onBlur handler when a blur occurs', () => {

            fireEvent.blur(inputElement)

            expect(onBlur).toHaveBeenCalled()
        })
    })

    async function rerenderComponent(
        options: {
            hidden?: boolean,
            accept?: string,
            capture?: boolean | 'user' | 'environment' | undefined,
            step?: string,
        } = {hidden: false, accept: '*'},
    ) {
        await rendering.rerender({
            id: inputId,
            label: inputLabel,
            type: inputType,
            placeholder: inputPlaceHolder,
            value: initialInputValue,
            hidden: options.hidden,
            accept: options.accept,
            capture: options.capture,
            step: options.step,
        })
    }

    afterEach(() => {
        rendering.unmount()
        cleanup()
        vi.clearAllMocks()
    })
})
