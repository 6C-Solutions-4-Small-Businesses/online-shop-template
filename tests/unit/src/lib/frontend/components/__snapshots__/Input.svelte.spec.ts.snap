// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input Component > html structure > should have match snapshot 1`] = `
<div>
  <div
    class="bg-red-500"
    data-testid="input-container"
  >
    <h4
      class="mb-1 text-black"
      data-testid="input-label-id"
    >
      input-label
       
    </h4>
     
    <input
      accept="*"
      capture="user"
      class="w-full h-fit  p-3.5 rounded-md border-gray-500 bg-white focus:outline-none focus:border-blue-500"
      data-testid="test-id"
      id="test-id"
      placeholder="input-placeholder"
      type="text"
    />
  </div>
</div>
`;
